edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/routing/CloudEndpoint.cs

```diff
@@ -150,27 +160,59 @@ public Task CloseAsync(CancellationToken token)
 
             bool IsTransientException(Exception ex) => ex is EdgeHubIOException || ex is EdgeHubConnectionException;
 
-            Util.Option<ICloudProxy> GetCloudProxy(IRoutingMessage routingMessage)
+            Util.Option<string> GetIdentity(IRoutingMessage routingMessage)
             {
                 if (routingMessage.SystemProperties.TryGetValue(SystemProperties.ConnectionDeviceId, out string deviceId))
                 {
-                    string id = routingMessage.SystemProperties.TryGetValue(SystemProperties.ConnectionModuleId, out string moduleId)
+                    return Option.Some(routingMessage.SystemProperties.TryGetValue(SystemProperties.ConnectionModuleId, out string moduleId)
                         ? $"{deviceId}/{moduleId}"
-                        : deviceId;
-                    Util.Option<ICloudProxy> cloudProxy = this.cloudEndpoint.cloudProxyGetterFunc(id);                        
-                    if(!cloudProxy.HasValue)
+                        : deviceId);
+                }
+                Events.DeviceIdNotFound(routingMessage);
+                return Option.None<string>();
+            }
+
+            Util.Option<ICloudProxy> GetCloudProxy(IRoutingMessage routingMessage)
+            {
+                return this.GetIdentity(routingMessage).Match(id =>
+                {
+                    Util.Option<ICloudProxy> cloudProxy = this.cloudEndpoint.cloudProxyGetterFunc(id);
+                    if (!cloudProxy.HasValue)
                     {
                         Events.IoTHubNotConnected(id);
                     }
                     return cloudProxy;
-                }
-                Events.DeviceIdNotFound(routingMessage);
-                return Option.None<ICloudProxy>();
+                }, () => Option.None<ICloudProxy>());
             }
 
             static bool IsRetryable(Exception ex) => ex != null && RetryableExceptions.Contains(ex.GetType());
         }
 
+        static class Metrics
+        {
+            static readonly CounterOptions EdgeHub2CMessageCountOptions = new CounterOptions
+            {
+                Name = "EdgeHub2CMessageSentCount",
```

Comment:
[super nit] is it possible to name these in snake_case, instead of PascalCase?

We should also make sure that any units are accounted for. This isn't a problem for counts but latency may need a unit. Here's the list of metrics that finagle collects as an example: https://twitter.github.io/finagle/guide/Metrics.html

Reply:
+1 for snake case..

Reply:
I will change to snake case, but what is the rationale for it?

Reply:
I personally like it the way it is.